[TensorExpr] Insert allocations for temporary buffer at the innermost valid scope.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#36836 [TensorExpr] Insert allocations for temporary buffer at the innermost valid scope.**

Differential Revision: [D21099913](https://our.internmc.facebook.com/intern/diff/D21099913)

Commit:
[TensorExpr] Insert allocations for temporary buffer at the innermost valid scope.

[ghstack-poisoned]

zheng-xq commented on torch/csrc/jit/tensorexpr/loopnest.cpp:
```diff
@@ -759,6 +759,97 @@ void LoopNest::computeInlineWithRandom(Stmt* s) {
+  return nullptr;
+}
+
+Block* findLowestContainingBlock(const std::vector<BufUse>& uses) {
```
Minor: add a TODO to see if we should improve the performance of this block for larger programs. Many systems I knew look at all the parents in reverse order, and find the lowest one.

Commit:
Update on "[TensorExpr] Insert allocations for temporary buffer at the innermost valid scope."


Differential Revision: [D21099913](https://our.internmc.facebook.com/intern/diff/D21099913)

[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/tensorexpr/loopnest.h:
```diff
@@ -68,6 +68,19 @@ class TORCH_API LoopNest {
 
 TORCH_API Stmt* FlattenIndexes(Stmt* s);
 
+struct BufUse {
```
I would defer this until we decide on how our dep-analysis should look like. I added the corresponding TODO comment.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/loopnest.cpp:
```diff
@@ -759,6 +759,97 @@ void LoopNest::computeInlineWithRandom(Stmt* s) {
+
+ private:
+  void visit(const Store* v) override {
+    if (stores_[v->buf()].insert(last_stmt_).second) {
```
It's actually an idiom I've seen and used a lot previously.
